src/GitVersionTask/GitVersionTaskBase.cs

```diff
@@ -1,72 +1,141 @@
+        {
+            switch (language)
+            {
+                case "C#":
+                    return "cs";
 
-        [Required]
-        public string SolutionDirectory { get; set; }
+                case "F#":
+                    return "fs";
 
-        public bool NoFetch { get; set; }
+                case "VB":
+                    return "vb";
 
-        public void LogDebug(string message)
-        {
-            BuildEngine.LogMessageEvent(new BuildMessageEventArgs(message, string.Empty, "GitVersionTask", MessageImportance.Low));
+                default:
+                    throw new Exception($"Unknown language detected: '{language}'");
+            }
         }
 
-        public void LogWarning(string message)
+        public static FileWriteInfo GetWorkingDirectoryAndFileNameAndExtension(
+            this String intermediateOutputPath,
+            String language,
+            String projectFile,
+            Func<String, String, String> fileNameWithIntermediatePath,
+            Func<String, String, String> fileNameNoIntermediatePath
+            )
         {
-            BuildEngine.LogWarningEvent(new BuildWarningEventArgs(string.Empty, string.Empty, null, 0, 0, 0, 0, message, string.Empty, "GitVersionTask"));
+            var fileExtension = language.GetFileExtension();
+            String workingDirectory, fileName;
+            if (intermediateOutputPath == null)
+            {
+                fileName = fileNameWithIntermediatePath(projectFile, fileExtension);
+                workingDirectory = TempFileTracker.TempPath;
+            }
+            else
+            {
+                workingDirectory = intermediateOutputPath;
+                fileName = fileNameNoIntermediatePath(projectFile, fileExtension);
+            }
+            return new FileWriteInfo(workingDirectory, fileName, fileExtension);
         }
 
-        public void LogInfo(string message)
+        public abstract class AbstractInput
         {
-            BuildEngine.LogMessageEvent(new BuildMessageEventArgs(message, string.Empty, "GitVersionTask", MessageImportance.Normal));
+            public String SolutionDirectory { get; set; }
+
+            public Boolean NoFetch { get; set; }
+
+            public virtual Boolean ValidateInput()
+            {
+                return !String.IsNullOrEmpty(this.SolutionDirectory);
+            }
         }
 
-        public void LogError(string message, string file = null)
+        public abstract class InputWithCommonAdditionalProperties : AbstractInput
         {
-            BuildEngine.LogErrorEvent(new BuildErrorEventArgs(string.Empty, string.Empty, file, 0, 0, 0, 0, message, string.Empty, "GitVersionTask"));
+            public String ProjectFile { get; set; }
+
+            public String IntermediateOutputPath { get; set; }
+
+            public String Language { get; set; }
+
+            public override Boolean ValidateInput()
```

Comment:
I think `void Validate()` throwing an exception would be better.

```suggestion
            public override void Validate()
```